atdgen/src/obuckle_emit.ml

```diff
@@ -101,16 +102,15 @@ let rec get_reader_name
   | Tvar (_, s) -> "read_" ^ Ox_emit.name_of_var s
 
   | Name (_, s, args, None, None) ->
-      let l = List.map (get_reader_name ~paren:true p) args in
-      let s = String.concat " " (name_f s :: l) in
-      if paren && l <> [] then "(" ^ s ^ ")"
-      else s
+      let l = List.map (get_reader_name ?type_annot ~paren:true p) args in
+      let s = String.concat " " (name_f (Printf.sprintf "%s: %s" s (type_annot_str type_annot)) :: l) in
+      "(" ^ s ^ ")"
```

Comment:
My idea was to add the type annotations to the variant constructors themselves. E.g `(A : variant)`. You can simply do this when printing those constructors (line 207 for the no arguments case for example). Btw, you shouldn't use `type_annot_str` here. We should just avoid printing a type annotation when it's not available. Otherwise it just adds noise.